src/x/cache/lru_cache.go

```diff
@@ -0,0 +1,528 @@
+	return e.Err.Error()
+}
+
+// Unwrap unwraps the underlying error
+func (e CachedError) Unwrap() error {
+	return e.Err
+}
+
+// As returns true if the caller is asking for the error as an cached error
+func (e CachedError) As(target interface{}) bool {
+	if cached, ok := target.(*CachedError); ok {
+		cached.Err = e.Err
+		return true
+	}
+	return false
+}
+
+var (
+	_ error = UncachedError{}
+	_ error = CachedError{}
+)
+
+var (
+	// ErrEntryNotFound is returned if a cache entry cannot be found.
+	ErrEntryNotFound = status.Error(codes.NotFound, "not found")
+
+	// ErrCacheFull is returned if we need to load an entry, but the cache is already full of entries that are loading.
+	ErrCacheFull = status.Error(codes.ResourceExhausted, "try again later")
+)
+
+// LRUOptions are the options to an LRU cache.
+type LRUOptions struct {
+	TTL                  time.Duration
+	InitialSize          int
+	MaxEntries           int
+	MaxConcurrency       int
+	CacheErrorsByDefault bool
+	Metrics              tally.Scope
+	Now                  func() time.Time
+}
+
+// LRU is a fixed size LRU cache supporting expiration, loading of entries that
+// do not exist, ability to cache negative results (e.g errors from load), and a
+// mechanism for preventing multiple goroutines from entering  the loader function
+// simultaneously for the same key, and a mechanism for restricting the amount of
+// total concurrency in the loader function.
+type LRU struct {
+	// TODO(mmihic): Consider striping these mutexes + the map entry so writes only
+	// take a lock out on a subset of the cache
+	mut               sync.Mutex
+	metrics           *lruCacheMetrics
+	cacheErrors       bool
+	maxEntries        int
+	ttl               time.Duration
+	concurrencyLeases chan struct{}
+	now               func() time.Time
+	byAccessTime      *list.List
+	byLoadTime        *list.List
+	entries           map[string]*lruCacheEntry
+}
+
+// NewLRU returns a new LRU with the provided options.
+func NewLRU(opts *LRUOptions) *LRU {
+	if opts == nil {
+		opts = &LRUOptions{}
+	}
+
+	ttl := opts.TTL
+	if ttl == 0 {
+		ttl = DefaultTTL
+	}
+
+	maxEntries := opts.MaxEntries
+	if maxEntries <= 0 {
+		maxEntries = DefaultMaxEntries
+	}
+
+	initialSize := opts.InitialSize
+	if initialSize <= 0 {
+		initialSize = int(math.Min(1000, float64(maxEntries)))
+	}
+
+	tallyScope := opts.Metrics
+	if tallyScope == nil {
+		tallyScope = tally.NoopScope
+	}
+
+	now := opts.Now
+	if now == nil {
+		now = time.Now
+	}
+
+	var concurrencyLeases chan struct{}
+	if opts.MaxConcurrency > 0 {
+		concurrencyLeases = make(chan struct{}, opts.MaxConcurrency)
+		for i := 0; i < opts.MaxConcurrency; i++ {
+			concurrencyLeases <- struct{}{}
+		}
+	}
+
+	return &LRU{
+		ttl:               ttl,
+		now:               now,
+		maxEntries:        maxEntries,
+		cacheErrors:       opts.CacheErrorsByDefault,
+		concurrencyLeases: concurrencyLeases,
+		metrics: &lruCacheMetrics{
+			entries:       tallyScope.Gauge(entriesGauge),
```

Comment:
nit: may be better to have `tallyScope` be scoped to `cache` rather than requiring one that's scoped to be passed in

Reply:
Yeah true, done 👍